Add ConfigMap support for seccomp custom profiles

pjbgf opened the pull request:
This proposal focuses on allowing users to load user-defined seccomp profiles from `ConfigMap` objects instead of loading them from physical files placed on nodes.

Feature: #135

For context:
https://github.com/kubernetes/kubernetes/issues/52827
https://github.com/kubernetes/kubernetes/issues/20870


/sig-node
/sig-auth

/priority important-longterm

/assign @tallclair
/cc @jessfraz

dchen1107 commented:
We discussed this proposal at today's SIG Node meeting. It is a usability improve for seccomp feature and would like to bundle it together with seccomp GA in v1.18 even there is no strict dependencies.

tallclair commented:
I haven't had a chance to look through the details yet, but I think it would be helpful to add a section (maybe under alternatives) justifying the use of a ConfigMap over a CRD. There's a good list of when each is appropriate here: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#should-i-use-a-configmap-or-a-custom-resource, and so noting which of those bullets apply in this situation would be helpful.

pjbgf commented:
@tallclair I have added the reasoning for choosing `ConfigMap` vs CRDs. PTAL

tallclair commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+- There is an existing, well-documented config file format.
+- The entire config file will be stored into one key of a configMap.
+- The main use of the config file is for the cluster to consume the file to pass it on to CRIs.
+- Users could potentially perform rolling updates via Deployment, when the file is updated.
```
Can you explain this more? I don't see how it's relevant to the CRD vs ConfigMap question?

pjbgf replied:
Sorry, my bad on that one, I misinterpreted what we needed there. I made some changes which should have improved this.

tallclair commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+- The entire config file will be stored into one key of a configMap.
+- The main use of the config file is for the cluster to consume the file to pass it on to CRIs.
+- Users could potentially perform rolling updates via Deployment, when the file is updated.
+
```
Some counter arguments:

- The configmap won't be mounted into a container, so it won't be reusing any Kubelet code.
- A CRD could be cluster-scoped (although I'm on the fence about whether that's actually desirable)
- A configmap can have multiple keys, which might not make sense in this case.
- A dedicated type could be immutable, to avoid some uncertainty around updates.
- A fully-structured CRD could provide static type checking for early failure (although this creates the undesirable tight coupling to the OCI format)

pjbgf replied:
Added all but one. I will explain better on the next comment.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+
+
+
+## Design Details
```
I have added this to the next batch of changes and updated Graduation Criteria accordingly.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+
+### Version Skew Strategy
+
+The new configmap based profiles will only be supported from this version on. Users will need to ensure that pods using the new feature won't be scheduled on nodes running older Kubelet versions.
```
I went through this and realised that the skew version validation is already in place. I am amending the KEP to contain the error messages for when either the API or Kubelet does not support it.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+
+#### Rollout of profile changes
+
+Seccomp profiles are applied at container creation time, therefore updating an existing user-defined profile will not cause any changes to the running containers that are using it. 
```
That's right. Any child process from the container will be bound by the seccomp profile without needing to reload it from the file.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+#### Rollout of profile changes
+
+Seccomp profiles are applied at container creation time, therefore updating an existing user-defined profile will not cause any changes to the running containers that are using it. 
+They will need to be restarted in order for the changes to take effect, which users will have to do manually. 
```
Just added a paragraph to that effect.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+
+An error message should be returned stating the profile could not be found. Containers targeted for seccomp profiles should not be able to start if their profiles are deleted. 
+
+If a running container has its policy deleted, nothing will happen until it has to be restarted, which in this case would lead to a failure as stated above.
```
With this proposal, the profiles would never be on disk. The logic would fetch it from `ConfigMap` and pass it straight to the OCI. Similarly to the current implementation we have for localhost:

```golang
fname := strings.TrimPrefix(seccompProfile, "localhost/")
file, err := ioutil.ReadFile(filepath.FromSlash(fname))
...
b := bytes.NewBuffer(nil)
json.Compact(b, file)

return []dockerOpt{{"seccomp", b.String(), msg}}, nil
```
https://github.com/pjbgf/kubernetes/blob/7e21146096e632dae18d0eeea8a2cd9eabf902a2/pkg/kubelet/dockershim/helpers_linux.go#L82

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+metadata:
+  name: my-custom-profile
+data:
+  profile.json: |-
```
@tallclair, thank you for raising those points. I tried to clarify them on the KEP.

The current proposal is that the Kubelet will be validating that the `ConfigMap` exists within the same namespace, containing a valid `json` file named `profile.json`. 
But I am not sure whether hard-coding the filename is the best way about it. Any suggestions?

pjbgf replied:
I put some more thought into it, and changed the profile name to be: `configmap/<configmap-name>/<filename>`. This removes the "magic" of having a fixed file name whilst giving users more flexibility.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,258 @@
+
+
+
+## Design Details
```
Added more information on that section. But basically it will solely be in-memory before it goes on to `runC`.

pjbgf commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
````diff
@@ -0,0 +1,258 @@
+  }
+``` 
+
+And the profile would be referenced as `configmap/my-custom-profile`.
````
Thank you again, with the changes I made it should be a lot clearer now.

pjbgf commented:
@tallclair PTAL

palnabarun commented:
/remove-lifecycle stale

palnabarun commented:
/milestone clear

saschagrunert commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,351 @@
+
+### No persistence to disk
+
+The profiles will not persisted to disk. The Kubelet will fetch the contents from `ConfigMap` and then returned to be later pass on to the OCI, as per code below:
```
```suggestion
The profiles will not persisted to disk. The Kubelet will fetch the contents from `ConfigMap` and then returned to be later pass on to the CRI, as per code below:
```

Do we really want to pass the whole JSON blob via the CRI? Seccomp profiles can be large (multiple kilobytes) from my point of view.

I'm also missing the scope what the container runtime should do with the profile information. If the whole profile gets passed via the CRI every time a workload starts, then we do not have to persist the profile state on disk, right?

pjbgf replied:
@saschagrunert Seccomp profiles can indeed grow large, depending on their targeted workload. The motivation to pass them in memory were primarily to have a simple implementation - like the one already in place for dockershim - whilst still overcoming the need to manage files or keep them in sync across nodes.

> I'm also missing the scope what the container runtime should do with the profile information. If the whole profile gets passed via the CRI every time a workload starts, then we do not have to persist the profile state on disk, right?

The initial idea was to send the profile on CreateContainerRequest and RunPodSandbox, which is when/home the filePath is currently being sent for "localhost/" seccomps.

saschagrunert replied:
Okay I see your arguments and will answer on the main thread. Let's close this discussion for now.

saschagrunert commented on keps/sig-node/20191002-seccomp-custom-profiles.md:
```diff
@@ -0,0 +1,351 @@
+
+## Alternatives
+
+### Use CRD instead of ConfigMap
+
+The decision to use `ConfigMap` was to avoid unnecessary complexity. However, this can be later reviewed through the Graduation process. The reasons considered whilst assessing the two different approaches were:
```
Honestly spoken I'm totally in favor of a CRD. I don't think that users write JSON seccomp profiles by hand. It's not possible to validate them upon creation and we cannot guarantee that a typo in a profile opens-up security issues.

evrardjp replied:
I agree with @saschagrunert .

pjbgf replied:
@saschagrunert @evrardjp I would also prefer the CRD approach, but ended up going for the ConfigMap to minimise changes needed so we could push this over the line sooner. 

Once the other details are ironed out, we could potentially consider moving to a CRD for Beta/GA. What do you think?

saschagrunert replied:
Sounds like a plan. I think we can iterate on this and keep the CRD out of scope for now (see my other comments)

evrardjp commented:
I am on the fence. This proposal is indeed more convient, and simple for users. However, I have to admit that passing big blobs to container runtimes for each container doesn't sound right to me.

The file based approach is simple, and just require configuration management.
I am pretty sure a few of those already exist, so we don't need to do this :p

pjbgf commented:
@saschagrunert @evrardjp thank you for taking the time reviewing the KEP. 

I don't mind changing the direction of this KEP to go for a file-based approach (which is already implemented and used by "localhost/" profiles). But given that seccomp is a security feature, I believe we should mitigate (or have some protection against) "profile race tampering".

In a file-based approach, between the creation of the file (by the kubelet, based off the ConfigMap/CRD) and its consumption (by the container runtime), the profile can be tampered with effectively disabling its protections. The end-users would be oblivious to this, as the containers would still run. It may be even harder to diagnose if it happens on a single node within their cluster.

At present there is no such protection, but I believe this would be a valid "requirement" for GA. 
Implementing a tamper-resistant file based approach, could require further changes to the CRI (and the container run-times), as they would have to receive and validate the payload hashes.

On the other hand, passing the entire profile to the CRI removes such need. As a level of "tamper resistance" could be obtained at transport level, which is something important between the kubelet and CRI regardless of this feature and therefore would be off-scope of this KEP.

The "least-effort" compromise could be to go for the file-based approach, using the existing infrastructure and use file permissions to protect against this. Having this well-documented and working together with kubeadmin/kops/kubespray to provide the community with an automated way to ensure file permissions are up to par. However, that feels to me like the weakest link in the seccomp chain implementation.

Please do share your thoughts and potential workarounds. I am quite keen to get this progressed and pushed over the line sooner rather than later.

saschagrunert commented:
> @saschagrunert @evrardjp thank you for taking the time reviewing the KEP.
> 
> I don't mind changing the direction of this KEP to go for a file-based approach (which is already implemented and used by "localhost/" profiles). But given that seccomp is a security feature, I believe we should mitigate (or have some protection against) "profile race tampering".

I acknowledge your arguments and the only issue I see is answering the question on: _How much information we want to transfer via the CRI?_ We right now have a state in the CRI where some information are transferred directly and others via local paths. If we say we generally want to move away from referencing local paths then I'm totally fine with going your proposed way.

---

Another more general point: We'd like to support you with your efforts to make Kubernetes more easily secure-able. With taking the seccomp GA KEP as well as the [new proposed release cycle](https://github.com/kubernetes/sig-release/pull/1065) into account, do you think we could bring this feature with 1.19 (together)?

We're planning to work in parallel on a more high level [seccomp-operator](https://github.com/saschagrunert/seccomp-operator/blob/master/RFC.md) in a separate repository. The idea is to provide a much higher level abstraction for seccomp (via the CRD) and sync the state into either files (like now) or config-maps (like your proposed KEP). What do you think about that? I'd love to see us working together on the overall story. 

(Later on we could promote your KEP to GA by adding the CRD from the seccomp-operator project, but this will be a tough run. Shipping a CRD via a feature gate is right now an unresolved question in Kubernetes.)

pjbgf commented:
> Another more general point: We'd like to support you with your efforts to make Kubernetes more easily secure-able. With taking the seccomp GA KEP as well as the new proposed release cycle into account, do you think we could bring this feature with 1.19 (together)?

That's brilliant, I do think we can deliver this for **1.19**, probably with a few small changes (i.e. use file-based profiles). The tamper-resistance can be a KEP on its own to consider Seccomp Profiles and AppArmor, which would consider the CRI impact.

The seccomp operator sounds like a great idea and I would be keen to part take. Around the subject, but not as comprehensive as the operator, I got another [KEP](https://github.com/kubernetes/enhancements/pull/1257) open that would bring some Kubernetes native profiles types. And on a side note, I have a pet project that auto-generates seccomp profiles reasonably for some workloads. Putting this all together may result in a better end-user experience. 

I think it may be worth having a catch-up to consider the "big-picture" and decide what pieces of the  puzzle we shall focus first to achieve that. From mid-next week I will have plenty of extra time to focus on this. I will reach you out on slack.

saschagrunert commented:
> The seccomp operator sounds like a great idea and I would be keen to part take. Around the subject, but not as comprehensive as the operator, I got another [KEP](https://github.com/kubernetes/enhancements/pull/1257) open that would bring some Kubernetes native profiles types. And on a side note, I have a pet project that auto-generates seccomp profiles reasonably for some workloads. Putting this all together may result in a better end-user experience.

That sounds great, seccomp profile recording via OCI hooks is mentioned in the RFC of the operator too. I think this would be a very valuable addition in the future. :+1: 

> I think it may be worth having a catch-up to consider the "big-picture" and decide what pieces of the puzzle we shall focus first to achieve that. From mid-next week I will have plenty of extra time to focus on this. I will reach you out on slack.

Yes, I setup a scheduled Zoom meeting for the whole topic:

Time: Apr 30, 2020 02:00 PM Amsterdam, Berlin, Rome, Stockholm, Vienna
Meeting: https://zoom.us/j/91379118387

pjbgf commented:
> Yes, I setup a scheduled Zoom meeting for the whole topic, everyone if free to join:
> Time: Apr 30, 2020 02:00 PM Amsterdam, Berlin, Rome, Stockholm, Vienna
> Meeting: https://zoom.us/j/91379118387

cc @tallclair

pjbgf commented:
@saschagrunert I created a [google doc](https://docs.google.com/document/d/1tBtsAFeR2FiqTfFXSiLUFWU76_vjLEMxhxmmvBkkoDI/edit?usp=sharing) to set an initial agenda. In it I also added a mind-map for me to talk through my understanding of the current state/outstanding work. I added a point for you to talk a bit about the seccomp-operator, feel free to add more stuff to it.

pjbgf commented:
This KEP is being withdrawn in favour of the [seccomp-operator](https://github.com/saschagrunert/seccomp-operator/blob/master/RFC.md), which will eventually bring similar features out of tree.
